include/ofi_mr.h

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2017 Intel Corporation, Inc. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif /* HAVE_CONFIG_H */
+
+#include "fi.h"
+#include "fi_util.h"
+
+/*
+ * SMRN - shared memory notification monitor
+ */
+
+struct util_smrn_attr {
+	struct fi_provider	*prov;
+
+	int (*init)(void);
+	int (*open)(void **notifier);
+	int (*close)(void *notifier);
+	int (*monitor)(void *notifier, void *addr,
+		       uint64_t len, uint64_t cookie);
+	int (*unmonitor)(void *notifier, uint64_t cookie);
+	int (*get_event)(void *notifier, void *buf, size_t len);
+};
+struct util_smrn {
+	fastlock_t		lock;
+	struct util_smrn_attr	attr;
+	void			*notifier;
+	struct dlist_entry	rq_head;
+	int			references;
+};
+
+struct util_smrn_rq {
+	fastlock_t		lock;
+	struct dlist_entry	list;
+	struct dlist_entry	entry;
+};
+
+struct util_smrn_context {
+	struct util_smrn_rq	*rq;
+	uint64_t		cookie;
```

Comment:
The cookie is necessary. It can be a pointer rather than a uint64_t, but the cookie refers to the actual memory registration cached by the memory registration cache. This provides a way for the consumer of SMRN events to retrieve the original pointer provided at registration time.